File: src/lib/services/booking.ts
import { SupabaseClient } from '@supabase/supabase-js';
import {
  BOOKING_FREE_MINS,
  BOOKING_FREE_MINS_SUBSCRIBED,
  PENALTY_HOURS,
  MAX_ACTIVE_BOOKINGS,
  MAX_ACTIVE_SESSIONS,
} from '../config';
import type { DbBooking } from '../types';

export class BookingService {
  constructor(private db: SupabaseClient) {}

  async createBooking(userId: string, cellId: string): Promise<DbBooking> {
    const { data: user } = await this.db
      .from('users')
      .select('*')
      .eq('id', userId)
      .maybeSingle();
    if (!user) throw new Error('User not found');
    if (user.has_debt) throw new Error(`User has debt: ${user.debt_amount}`);

    const { count: activeBookings } = await this.db
      .from('bookings')
      .select('*', { count: 'exact', head: true })
      .eq('user_id', userId)
      .eq('status', 'ACTIVE');
    if ((activeBookings ?? 0) >= MAX_ACTIVE_BOOKINGS) throw new Error('Max active bookings reached');

    const { count: activeSessions } = await this.db
      .from('sessions')
      .select('*', { count: 'exact', head: true })
      .eq('user_id', userId)
      .is('end_at', null);
    if ((activeSessions ?? 0) >= MAX_ACTIVE_SESSIONS) throw new Error('Max active sessions reached');

    const { data: cell } = await this.db
      .from('cells')
      .select('*')
      .eq('id', cellId)
      .maybeSingle();
    if (!cell || cell.status !== 'FREE') throw new Error('Cell is not available');

    const now = new Date();
    const { data: penaltyBooking } = await this.db
      .from('bookings')
      .select('id')
      .eq('user_id', userId)
      .in('status', ['CANCELLED', 'EXPIRED'])
      .gt('penalty_until', now.toISOString())
      .maybeSingle();
    const isFree = penaltyBooking === null;

    const { data: sub } = await this.db
      .from('subscriptions')
      .select('id')
      .eq('user_id', userId)
      .eq('is_active', true)
      .gt('end_at', now.toISOString())
      .maybeSingle();
    const freeMins = sub ? BOOKING_FREE_MINS_SUBSCRIBED : BOOKING_FREE_MINS;
    const endsAt = new Date(now.getTime() + freeMins * 60 * 1000);

    const { data: booking, error } = await this.db
      .from('bookings')
      .insert({
        user_id: userId,
        cell_id: cellId,
        status: 'ACTIVE',
        is_free: isFree,
        ends_at: endsAt.toISOString(),
      })
      .select()
      .single<DbBooking>();
    if (error) throw new Error(error.message);

    await this.db.from('cells').update({ status: 'BUSY' }).eq('id', cellId);
    return booking;
  }

  async cancelBooking(bookingId: string, userId: string): Promise<Date> {
    const { data: booking } = await this.db
      .from('bookings')
      .select('*')
      .eq('id', bookingId)
      .maybeSingle<DbBooking>();
    if (!booking || booking.user_id !== userId) throw new Error('Booking not found');
    if (booking.status !== 'ACTIVE') throw new Error('Booking is not active');

    const now = new Date();
    const penaltyUntil = new Date(now.getTime() + PENALTY_HOURS * 60 * 60 * 1000);

    await this.db
      .from('bookings')
      .update({ status: 'CANCELLED', penalty_until: penaltyUntil.toISOString() })
      .eq('id', bookingId);
    await this.db.from('cells').update({ status: 'FREE' }).eq('id', booking.cell_id);
    return penaltyUntil;
  }

  async expireBooking(bookingId: string): Promise<DbBooking | null> {
    const { data: booking } = await this.db
      .from('bookings')
      .select('*')
      .eq('id', bookingId)
      .maybeSingle<DbBooking>();
    if (!booking || booking.status !== 'ACTIVE') return null;

    const now = new Date();
    if (new Date(booking.ends_at) > now) return null;

    await this.db.from('bookings').update({ status: 'EXPIRED' }).eq('id', bookingId);

    const { data: activeSession } = await this.db
      .from('sessions')
      .select('id')
      .eq('booking_id', bookingId)
      .is('end_at', null)
      .maybeSingle();
    if (!activeSession) {
      await this.db.from('cells').update({ status: 'FREE' }).eq('id', booking.cell_id);
    }
    return booking;
  }
}

File: backend/app/services/booking.py
from datetime import datetime, timedelta, timezone
from app.config import (
    BOOKING_FREE_MINS, BOOKING_FREE_MINS_SUBSCRIBED,
    PENALTY_HOURS, MAX_ACTIVE_BOOKINGS, MAX_ACTIVE_SESSIONS,
)


class BookingService:
    def __init__(self, db):
        self.db = db

    async def create_booking(self, user_id: str, cell_id: str):
        user = await self.db.user.find_unique(where={"id": user_id})
        if not user:
            raise ValueError("User not found")
        if user.hasDebt:
            raise ValueError(f"User has debt: {user.debtAmount}")

        active_bookings = await self.db.booking.count(
            where={"userId": user_id, "status": "ACTIVE"}
        )
        if active_bookings >= MAX_ACTIVE_BOOKINGS:
            raise ValueError("Max active bookings reached")

        active_sessions = await self.db.session.count(
            where={"userId": user_id, "endAt": None}
        )
        if active_sessions >= MAX_ACTIVE_SESSIONS:
            raise ValueError("Max active sessions reached")

        cell = await self.db.cell.find_unique(where={"id": cell_id})
        if not cell or cell.status != "FREE":
            raise ValueError("Cell is not available")

        # Check penalty window
        now = datetime.now(timezone.utc)
        penalty_booking = await self.db.booking.find_first(
            where={
                "userId": user_id,
                "penaltyUntil": {"gt": now},
                "status": {"in": ["CANCELLED", "EXPIRED"]},
            }
        )
        is_free = penalty_booking is None

        # Check subscription
        sub = await self.db.subscription.find_first(
            where={"userId": user_id, "isActive": True, "endAt": {"gt": now}}
        )
        free_mins = BOOKING_FREE_MINS_SUBSCRIBED if sub else BOOKING_FREE_MINS

        ends_at = now + timedelta(minutes=free_mins)

        booking = await self.db.booking.create(data={
            "userId": user_id,
            "cellId": cell_id,
            "status": "ACTIVE",
            "isFree": is_free,
            "endsAt": ends_at,
        })

        # Mark cell as busy
        await self.db.cell.update(where={"id": cell_id}, data={"status": "BUSY"})
        return booking

    async def cancel_booking(self, booking_id: str, user_id: str):
        booking = await self.db.booking.find_unique(where={"id": booking_id})
        if not booking or booking.userId != user_id:
            raise ValueError("Booking not found")
        if booking.status != "ACTIVE":
            raise ValueError("Booking is not active")

        now = datetime.now(timezone.utc)
        penalty_until = now + timedelta(hours=PENALTY_HOURS)

        await self.db.booking.update(
            where={"id": booking_id},
            data={"status": "CANCELLED", "penaltyUntil": penalty_until},
        )
        await self.db.cell.update(
            where={"id": booking.cellId},
            data={"status": "FREE"},
        )
        return penalty_until

    async def expire_booking(self, booking_id: str):
        booking = await self.db.booking.find_unique(where={"id": booking_id})
        if not booking or booking.status != "ACTIVE":
            return None

        now = datetime.now(timezone.utc)
        if booking.endsAt > now:
            return None  # Not yet expired

        await self.db.booking.update(
            where={"id": booking_id},
            data={"status": "EXPIRED"},
        )
        # Free cell if no active session started
        active_session = await self.db.session.find_first(
            where={"bookingId": booking_id, "endAt": None}
        )
        if not active_session:
            await self.db.cell.update(
                where={"id": booking.cellId},
                data={"status": "FREE"},
            )
        return booking
